src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/transport/action/addpolicy/TransportAddPolicyAction.kt

```diff
@@ -60,12 +81,36 @@ class TransportAddPolicyAction @Inject constructor(
         private val request: AddPolicyRequest
     ) {
         private lateinit var startTime: Instant
+        private val indicesToAdd = mutableMapOf<String, String>() // uuid: name
         private val failedIndices: MutableList<FailedIndex> = mutableListOf()
-        private val indicesToAddPolicyTo: MutableList<Index> = mutableListOf()
-        private var updated: Int = 0
 
-        @Suppress("SpreadOperator")
         fun start() {
+            ismIndices.checkAndUpdateIMConfigIndex(object : ActionListener<AcknowledgedResponse> {
+                override fun onResponse(response: AcknowledgedResponse) {
+                    onCreateMappingsResponse(response)
+                }
+
+                override fun onFailure(t: Exception) {
+                    actionListener.onFailure(t)
```

Comment:
Use ExceptionsHelper.unwrapCause on the exception.

Reply:
cleaned up all files with onFailure in indexstatemanagement